### DISCUSSION TITLE
Is there a way to code a text in bash and send to target computer using the OTG keyboard

### DISCUSSION THREAD

**Discussion by wilsonnkwan** [tiny-pilot/tinypilot#826] [Q&A] [answered]
Hi, I am writing to check with you if we can manually code some text like "Hello World" using bash and use cron to activate that script and send the input into the OTG USB Keyboard to target computer.

This is particularly helpful when I want to type something repetitive but I don't want to store it in the target computer. Hope you can help. Thanks :)

**Comment by mtlynch** [comment, marked as answer]
It's possible, though we don't have great tooling around it in the community version.

From bash, you can run commands like the following to type "Hi" on the target system's keyboard:

```bash
# H (Right shift + h)
echo -ne "\x20\0\xb\0\0\0\0\0" > /dev/hidg0
# i
echo -ne "\0\0\xc\0\0\0\0\0" > /dev/hidg0
# Release all keys
echo -ne "\0\0\0\0\0\0\0\0" > /dev/hidg0
```

For more details, see: https://mtlynch.io/key-mime-pi/#usb-device-emulation

If you're interested in an Enterprise license, it includes a [REST API](https://tinypilotkvm.com/api/v1) for sending keystrokes programmatically.

**Reply by wilsonnkwan** [reply]
Hi, thanks. Supposed that I want to code the keystroke, Ctrl+Shift+S, how should i do about doing it?

How can I do that? Thanks in advanced :)

**Reply by mtlynch** [reply]
If you run TinyPilot in debug mode, it will show you what HID codes it's sending to the target computer, so you could type the keys you want, record the HID codes, and then put them into a bash script.

There are two ways to run TinyPilot in debug mode. You can [set up a dev environment](https://github.com/tiny-pilot/tinypilot/blob/master/CONTRIBUTING.md#setting-up-a-development-environment) on a desktop system. If you check the logs as the dev instance is running, you'll see log messages like this:

```
[2021-11-24 16:25:00,737] DEBUG in socket_api: received keystroke message: {'metaLeft': False, 'metaRight': False, 'altLeft': False, 'altRight': False, 'shiftLeft': False, 'shiftRight': False, 'ctrlLeft': False, 'ctrlRight': False, 'key': 'Enter', 'code': 'Enter'}
[2021-11-24 16:25:00,737] DEBUG in write: writing to HID interface /dev/null: 0x00 0x00 0x28 0x00 0x00 0x00 0x00 0x00
[2021-11-24 16:25:00,742] DEBUG in write: writing to HID interface /dev/null: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
```

The `/dev/null` lines show the HID codes.

The other way to run in debug mode is to modify the TinyPilot installation on your Pi by running `sudo nano /lib/systemd/system/tinypilot.service` and then where you see the `Environment=` lines, add a line that says `Environment=DEBUG=1`. Then save the file and run `sudo systemctl daemon-reload && sudo service tinypilot restart`. After that, when you check your logs with System > View Logs, you'll see the debug lines with the HID codes, as above.
